firewall/master.go

```diff
@@ -31,91 +31,170 @@ import (
 // 4. DecideOnLink
 //		is called when when the first packet of a link arrives only if communication has verdict UNDECIDED or CANTSAY
 
-// DecideOnCommunicationBeforeDNS makes a decision about a communication before the dns query is resolved and intel is gathered.
-func DecideOnCommunicationBeforeDNS(comm *network.Communication) {
+// DecideOnConnection makes a decision about a connection.
+func DecideOnConnection(conn *network.Connection, pkt packet.Packet) {
 	// update profiles and check if communication needs reevaluation
-	if comm.UpdateAndCheck() {
-		log.Infof("firewall: re-evaluating verdict on %s", comm)
-		comm.ResetVerdict()
+	if conn.UpdateAndCheck() {
+		log.Infof("filter: re-evaluating verdict on %s", conn)
+		conn.Verdict = network.VerdictUndecided
 	}
 
-	// check if need to run
-	if comm.GetVerdict() != network.VerdictUndecided {
+	// grant self
+	if conn.Process().Pid == os.Getpid() {
+		log.Infof("filter: granting own connection %s", conn)
+		conn.Verdict = network.VerdictAccept
 		return
 	}
 
-	// grant self
-	if comm.Process().Pid == os.Getpid() {
-		log.Infof("firewall: granting own communication %s", comm)
-		comm.Accept("")
-		return
+	// check if process is communicating with itself
+	if pkt != nil {
+		if conn.Process().Pid >= 0 && pkt.Info().Src.Equal(pkt.Info().Dst) {
```

Comment:
What's that `conn.Process().Pid >= 0` check for? Btw, what about process A listening on `127.0.0.100` (localhost) and communicating with itself using a new socket. Src would be 127.0.0.1 and Dst 127.0.0.100?

Reply:
There is a `*Process` instance that is used for when the process could not be identified - an "Unknown Process", which uses the Pid `-1`.

In the case of different localhost addresses this check would fail. I will add a TODO to investigate this later. I think the probability of this happening is very low though.